sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Group.java

```diff
@@ -0,0 +1,684 @@
+ * schema, and generating an output schema based on these aggregations. For example:
+ *
+ * <pre>{@code
+ * {@literal PCollection<KV<Row, Row>>} aggregated = purchases
+ *      .apply(Group.byFieldNames("userId', "country")
+ *          .aggregateField("cost", Sum.ofLongs(), "total_cost")
+ *          .aggregateField("cost", Top.<Long>largestLongsFn(10), "top_purchases")
+ *          .aggregateField("cost", ApproximateQuantilesCombineFn.create(21),
+ *            Field.of("transactionDurations", FieldType.array(FieldType.INT64)));
+ * }</pre>
+ *
+ * <p>The result will be a new row schema containing the fields total_cost, top_purchases, and
+ * transactionDurations, containing the sum of all purchases costs (for that user and country), the
+ * top ten purchases, and a histogram of transaction durations.
+ *
+ * <p>Note that usually the field type can be automatically inferred from the {@link CombineFn}
+ * passed in. However sometimes it cannot be inferred, due to Java type erasure, in which case a
+ * {@link Field} object containing the field type must be passed in. This is currently the case for
+ * ApproximateQuantilesCombineFn in the above example.
+ */
+@Experimental(Kind.SCHEMAS)
+public class Group {
+  /**
+   * Returns a transform that groups all elements in the input {@link PCollection}. The returned
+   * transform contains further builder methods to control how the grouping is done.
+   */
+  public static <T> Global<T> globally() {
+    return new Global<>();
+  }
+
+  /**
+   * Returns a transform that groups all elements in the input {@link PCollection} keyed by the list
+   * of fields specified. The output of this transform will be a {@link KV} keyed by a {@link Row}
+   * containing the specified extracted fields. The returned transform contains further builder
+   * methods to control how the grouping is done.
+   */
+  public static <T> ByFields<T> byFieldNames(String... fieldNames) {
+    return new ByFields<>(FieldAccessDescriptor.withFieldNames(fieldNames));
+  }
+
+  /**
+   * Returns a transform that groups all elements in the input {@link PCollection} keyed by the list
+   * of fields specified. The output of this transform will be a {@link KV} keyed by a {@link Row}
+   * containing the specified extracted fields. The returned transform contains further builder
+   * methods to control how the grouping is done.
+   */
+  public static <T> ByFields<T> byFieldIds(Integer... fieldIds) {
+    return new ByFields<>(FieldAccessDescriptor.withFieldIds(fieldIds));
+  }
+
+  /**
+   * Returns a transform that groups all elements in the input {@link PCollection} keyed by the
+   * fields specified. The output of this transform will be a {@link KV} keyed by a {@link Row}
+   * containing the specified extracted fields. The returned transform contains further builder
+   * methods to control how the grouping is done.
+   */
+  public static <T> ByFields<T> byFieldAccessDescriptor(FieldAccessDescriptor fieldAccess) {
```

Comment:
Can we avoid making this public? At least mark as `@Internal`?

Reply:
This is the only way to group by a nested field today, so it needs to be public. Later if we support other ways of expressing nested fields (e.g. jsonpath) we can reconsider whether this needs to be public.